visual-js/visual-wdio/src/SauceVisualService.ts

```diff
@@ -163,86 +147,65 @@ ${url.padStart(100, ' ')}
 `;
 }
 
-function splitIgnorables(
-  a: Array<IgnoreRegion | WdioElement | WdioElement[]>,
-): { elements: WdioElement[]; regions: IgnoreRegion[] } {
-  const elements: WdioElement[] = [];
-  const regions: IgnoreRegion[] = [];
-  for (let i = 0; i < a.length; i++) {
-    const x = a[i];
-
-    if (Array.isArray(x)) {
-      elements.concat(...x);
-    } else if (isWdioElement(x)) {
-      elements.push(x);
-    } else {
-      try {
-        const region = validateIgnoreRegion(x);
-        regions.push(region);
-      } catch (e) {
-        throw new Error(
-          `Ignored element options.ignore[${i}] is not valid: ${e}`,
-        );
-      }
-    }
+const getDiffingOptions = (
+  region: RegionType,
+): DiffingOptionsIn | undefined => {
+  const { element: __, ...rest } = region;
+  if ('enableOnly' in rest || 'disableOnly' in rest) {
+    return selectiveRegionOptionsToDiffingOptions(rest);
   }
-  return { elements, regions };
-}
-
-async function wdio2IgnoreRegions(
-  elements: WdioElement[],
-): Promise<IgnoreRegion[]> {
-  // This translation is done for client side scripts by wdio
-  const asHtmlElements = elements as unknown as HTMLElement[];
-
-  const ignoreRegions = await browser.execute(
-    clientSideIgnoreRegionsFromElements,
-    asHtmlElements,
-  );
+  return undefined;
+};
 
-  if (ignoreRegions.length !== elements.length) {
-    throw new Error(
-      'Internal error while getting the bounding rect for elements',
-    );
-  }
+const parseRegionsForAPI = async (
+  regions: WdioSelectiveRegion[],
+  ignore: Ignorable[],
+): Promise<{
+  ignoreRegions: RegionIn[];
+  ignoreElements: ElementIn[];
+}> => {
+  const allRegions: RegionType[] = [
+    ...(regions ?? []),
+    ...(ignore ?? []).map((element) => ({ element })),
+  ];
+
+  const ignoreElements: ElementIn[] = [];
+  const ignoreRegions: RegionIn[] = [];
+
+  const pushElement = async (region: RegionType) => {
+    const { element: maybePromiseElement, ...other } = region;
+    const element = await maybePromiseElement;
+
+    if (Array.isArray(element)) {
+      for await (const subElement of element) {
+        await pushElement({
```

Comment:
I am not sure if the recursion is necessary: Are we dealing with a deep tree here? I thought it is depth 2, i.e. maps to `RegionIn[][]`.

I would also try an linearize the code, by converting to an array if necessary:

```ts
const ignoreRegions = Promise.all(ignoreElements.maps(async region => {
  const {element, ...other } = await region.element;
  const subElements = Array.isArray(element) ? element : [element];
  return subElements.map(sub => ({
    ...(isWdio(sub) ? {id: ....} : {...sub}),
    ...other,
    diffingOptions: getDiffingOptions(region),
  }))
})).flat()
```

Since this code is similar to in all JS frameworks, we might want to move it the `visual` package and only keep the detection and conversion of wdio elements in here. (Not sure about this one.)

Reply:
We'd just have to then re-filter those back out into separate arrays. 

The elements used by Nightwatch also vary. WDIO uses, well, WDIO, where Nightwatch AFAIK uses selenium underneath which has different element types. Though I suppose that could be exported and allowing `isWdio` (is webelement, maybe) to be customized.